File: leetcode/lc-us/cpp/2458.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}

    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}

    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

const int N = 1e5 + 10;

int mx;
int h[N];

class Solution {
    void dfs1(TreeNode *node, int depth) {
        h[node->val] = mx;
        mx = max(mx, depth);
        if (node->left) dfs1(node->left, depth + 1);
        if (node->right) dfs1(node->right, depth + 1);
    }

    void dfs2(TreeNode *node, int depth) {
        h[node->val] = max(h[node->val], mx);
        mx = max(mx, depth);
        if (node->right) dfs2(node->right, depth + 1);
        if (node->left) dfs2(node->left, depth + 1);
    }

public:
    vector<int> treeQueries(TreeNode *root, vector<int> &queries) {
        memset(h, 0, sizeof h);

        mx = 0, dfs1(root, 0);
        mx = 0, dfs2(root, 0);

        int n = (int) queries.size();
        vector<int> res(n);
        for (int i = 0; i < n; i++) {
            res[i] = h[queries[i]];
        }
        return res;
    }
};

File: leetcode/lc-us/py/2458.py
from collections import defaultdict

from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def __init__(self):
        self.mx = 0

    def treeQueries(self, root: Optional[TreeNode], queries: List[int]) -> List[int]:
        depth = defaultdict(int)

        def dfs1(node: 'TreeNode', d: int):
            depth[node.val] = self.mx
            self.mx = max(self.mx, d)
            if node.left:
                dfs1(node.left, d + 1)
            if node.right:
                dfs1(node.right, d + 1)

        def dfs2(node: 'TreeNode', d: int):
            depth[node.val] = max(depth[node.val], self.mx)
            self.mx = max(self.mx, d)
            if node.right:
                dfs2(node.right, d + 1)
            if node.left:
                dfs2(node.left, d + 1)

        dfs1(root, 0)
        self.mx = 0
        dfs2(root, 0)
        return [depth[q] for q in queries]
